hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/common/SubmissionManager.java

```diff
@@ -0,0 +1,114 @@
+
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.PLATFORM_TRANSACTION_NOT_CREATED;
+import static java.util.Objects.requireNonNull;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Parser;
+import com.hedera.node.app.service.mono.context.properties.NodeLocalProperties;
+import com.hedera.node.app.service.mono.context.properties.Profile;
+import com.hedera.node.app.service.mono.records.RecordCache;
+import com.hedera.node.app.service.mono.stats.MiscSpeedometers;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import com.swirlds.common.system.Platform;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.ByteBuffer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** The {@code SubmissionManager} provides functionality to submit transactions to the platform. */
+public class SubmissionManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(SubmissionManager.class);
+
+    private final Platform platform;
+    private final RecordCache recordCache;
+    private final NodeLocalProperties nodeLocalProperties;
+    private final MiscSpeedometers speedometers;
+
+    /**
+     * Constructor of {@code SubmissionManager}
+     *
+     * @param platform the {@link Platform} to which transactions will be submitted
+     * @param recordCache the {@link RecordCache} that tracks submitted transactions
+     * @param nodeLocalProperties the {@link NodeLocalProperties} that keep local properties
+     * @param speedometers metrics related to submissions
+     */
+    public SubmissionManager(
+            @NonNull final Platform platform,
+            @NonNull final RecordCache recordCache,
+            @NonNull final NodeLocalProperties nodeLocalProperties,
+            @NonNull final MiscSpeedometers speedometers) {
+        this.platform = requireNonNull(platform);
+        this.recordCache = requireNonNull(recordCache);
+        this.nodeLocalProperties = requireNonNull(nodeLocalProperties);
+        this.speedometers = requireNonNull(speedometers);
+    }
+
+    /**
+     * Submit a transaction to the {@link Platform}
+     *
+     * @param txBody the {@link TransactionBody} that should be submitted to the platform
+     * @param byteBuffer the {@link ByteBuffer} of the data that should be submitted
+     * @param parser the {@link Parser} that is used to eventually parse the {@link
+     *     TransactionBody#getUncheckedSubmit()}
+     * @throws NullPointerException if one of the arguments is {@code null}
+     * @throws PreCheckException if the transaction could not be submitted
+     */
+    public void submit(
+            @NonNull final TransactionBody txBody,
+            @NonNull final ByteBuffer byteBuffer,
+            @NonNull final Parser<TransactionBody> parser)
+            throws PreCheckException {
+        requireNonNull(txBody);
+        requireNonNull(byteBuffer);
+        requireNonNull(parser);
```

Comment:
Do we need `requireNonNull(...)` with the `@NonNull` annotations?

Reply:
Unfortunately, we need both. The annotation provides a hint to IDEs and code analyzers. But a developer can choose to ignore all warnings and still pass `null`. In that case we want to throw an NPE, which is done by `requireNonNull()`.